### ISSUE TITLE
6. Differentiable Symbolic DSL for Logic and Constraints

### ISSUE THREAD

**Issue by ricardoamartinez** [ricardoamartinez/koopman-physics-neural-cellular-automata#7]
# Differentiable Symbolic DSL for Logic and Constraints  

## Overview
Integrate a **differentiable Domain-Specific Language (DSL)** component that allows symbolic rules and programs to influence agent computations while maintaining full differentiability. This enables injection of high-level knowledge and constraints in a structured way, with gradients flowing from symbolic rules into neural network parameters.

### Core DSL Parser and Evaluator
Implement a simple interpreter that parses symbolic expressions and executes PyTorch operations:

```python
# Example DSL program representation (could be a list of tuples)
dsl_program = [
    ("goal", 5.0, 5.0),
    ("stay_inside", 0.0, 10.0, 0.0, 10.0),
    ("face_direction", 1.57),  # π/2 radians
    ("avoid_obstacle", 3.0, 3.0, 1.0)  # x, y, radius
]

def evaluate_dsl(latent, state):
    """
    Evaluate DSL program and return differentiable loss
    latent: the agent's internal latent state (torch tensor)
    state: the agent's actual or predicted physical state (e.g., position) as torch tensor
    Returns: total differentiable loss tensor
    """
    total_loss = torch.tensor(0.0, device=state.device)
    
    for instr in dsl_program:
        if instr[0] == "goal":
            _, goal_x, goal_y = instr
            # Suppose state contains position at indices [0,1]
            dx = state[0] - torch.tensor(goal_x, device=state.device)
            dy = state[1] - torch.tensor(goal_y, device=state.device)
            dist = torch.sqrt(dx*dx + dy*dy)
            # Add loss for distance to goal (to be minimized)
            total_loss = total_loss + dist
            
        elif instr[0] == "stay_inside":
            _, xmin, xmax, ymin, ymax = instr
            # Penalize being outside [xmin,xmax]x[ymin,ymax]
            px, py = state[0], state[1]
            penalty = torch.relu(px - torch.tensor(xmax, device=px.device))
            penalty += torch.relu(torch.tensor(xmin, device=px.device) - px)
            penalty += torch.relu(py - torch.tensor(ymax, device=py.device))
            penalty += torch.relu(torch.tensor(ymin, device=py.device) - py)
            total_loss = total_loss + 10.0 * penalty  # weight constraint higher
            
        elif instr[0] == "face_direction":
            _, target_theta = instr
            # Assuming state[3] contains current orientation
            current_theta = state[3]
            angle_diff = torch.abs(current_theta - torch.tensor(target_theta, device=current_theta.device))
            # Handle angle wrapping (shortest angular distance)
            angle_diff = torch.min(angle_diff, 2*torch.pi - angle_diff)
            total_loss = total_loss + angle_diff
            
        elif instr[0] == "avoid_obstacle":
            _, obs_x, obs_y, radius = instr
            # Penalize being too close to obstacle
            px, py = state[0], state[1]
            dist_to_obs = torch.sqrt((px - obs_x)**2 + (py - obs_y)**2)
            violation = torch.relu(radius - dist_to_obs)  # positive if inside obstacle
            total_loss = total_loss + 20.0 * violation  # high penalty for collision
            
    return total_loss
```

#### Conditional Logic with Soft Conditionals
Implement differentiable conditional statements:

```python
def soft_conditional(condition, then_value, else_value, sharpness=10.0):
    """
    Differentiable IF-THEN-ELSE using sigmoid
    condition: tensor that should be > 0 for "true"
    then_value: value to return if condition is true
    else_value: value to return if condition is false
    sharpness: controls how sharp the transition is
    """
    gate = torch.sigmoid(sharpness * condition)
    return gate * then_value + (1 - gate) * else_value

# Usage in DSL evaluation:
def evaluate_conditional_dsl(state):
    """Example: IF close_to_obstacle THEN turn_left ELSE move_forward"""
    # Check if close to obstacle
    obstacle_pos = torch.tensor([3.0, 3.0], device=state.device)
    distance = torch.norm(state[:2] - obstacle_pos)
    close_to_obstacle = 1.0 - distance  # positive when close
    
    # Define actions
    turn_penalty = torch.abs(state[3] - torch.pi/2)  # want to turn left (π/2)
    forward_penalty = torch.abs(state[4])  # want zero angular velocity (go straight)
    
    # Conditional penalty
    penalty = soft_conditional(close_to_obstacle, turn_penalty, forward_penalty)
    return penalty
```

#### Loop Constructs with Differentiable Iteration
Implement bounded loops for temporal reasoning:

```python
def differentiable_for_loop(initial_state, loop_body, num_iterations=5):
    """
    Differentiable loop that applies loop_body for num_iterations
    initial_state: starting state tensor
    loop_body: function that takes state and returns new state and loss
    """
    current_state = initial_state
    total_loop_loss = torch.tensor(0.0, device=initial_state.device)
    
    for i in range(num_iterations):
        # Apply loop body
        next_state, step_loss = loop_body(current_state, i)
        total_loop_loss += step_loss
        
        # Update state (can be learned or deterministic)
        current_state = next_state
    
    return current_state, total_loop_loss

# Example loop body: predict trajectory and penalize constraint violations
def trajectory_prediction_loop(state, step):
    """Predict next state and accumulate constraint violations"""
    # Simple dynamics model (learnable)
    velocity = state[3:5]  # assuming state = [x, y, theta, vx, vy]
    next_position = state[:2] + 0.1 * velocity  # dt = 0.1
    next_state = torch.cat([next_position, state[2:]])
    
    # Compute constraint violation at this step
    step_loss = evaluate_dsl(None, next_state)
    
    return next_state, step_loss
```

#### Strategy A: Loss Augmentation
Add DSL objectives to training loss:

```python
def compute_total_loss(model_output, target, state):
    """Combine model loss with DSL constraints"""
    # Primary model loss (e.g., JEPA, reconstruction, etc.)
    model_loss = F.mse_loss(model_output, target)
    
    # DSL constraint loss
    dsl_loss = evaluate_dsl(None, state)
    
    # Weighted combination
    lambda_dsl = 0.1  # weight for DSL constraints
    total_loss = model_loss + lambda_dsl * dsl_loss
    
    return total_loss
```

#### Strategy B: Action Optimization
Use DSL for active inference action selection:

```python
def active_inference_with_dsl(current_state, model_dynamics, action_dim=6, n_steps=10):
    """
    Optimize actions to minimize DSL objectives
    current_state: current agent state tensor
    model_dynamics: function that predicts next state given current state and action
    """
    # Initialize action parameters as learnable variables
    action_params = torch.randn(action_dim, requires_grad=True, device=current_state.device)
    action_optimizer = torch.optim.SGD([action_params], lr=0.1)
    
    for step in range(n_steps):
        # Bound actions to reasonable range
        bounded_action = torch.tanh(action_params)
        
        # Predict next state using internal model
        predicted_next_state = model_dynamics(current_state, bounded_action)
        
        # Evaluate DSL constraints on predicted state
        dsl_cost = evaluate_dsl(None, predicted_next_state)
        
        # Add small action regularization
        action_cost = 0.01 * torch.norm(bounded_action)
        
        # Total cost to minimize
        total_cost = dsl_cost + action_cost
        
        # Optimize action parameters
        action_optimizer.zero_grad()
        total_cost.backward()
        action_optimizer.step()
    
    # Return optimized action
    final_action = torch.tanh(action_params).detach()
    return final_action
```

### 1. DSL Program Representation
Define structured representation for symbolic programs:

```python
class DSLInstruction:
    def __init__(self, operation, *args):
        self.operation = operation
        self.args = args
    
    def __repr__(self):
        return f"{self.operation}({', '.join(map(str, self.args))})"

class DSLProgram:
    def __init__(self, instructions):
        self.instructions = [DSLInstruction(op, *args) if isinstance(op, str) 
                           else op for op, *args in instructions]
    
    def evaluate(self, latent, state):
        """Evaluate all instructions and return total loss"""
        total_loss = torch.tensor(0.0, device=state.device)
        
        for instr in self.instructions:
            loss = self.evaluate_instruction(instr, latent, state)
            total_loss += loss
            
        return total_loss
    
    def evaluate_instruction(self, instr, latent, state):
        """Evaluate single instruction"""
        if instr.operation == "goal":
            return self._goal_loss(state, *instr.args)
        elif instr.operation == "stay_inside":
            return self._boundary_loss(state, *instr.args)
        elif instr.operation == "avoid_obstacle":
            return self._obstacle_loss(state, *instr.args)
        # Add more operations as needed
        else:
            raise ValueError(f"Unknown operation: {instr.operation}")
```

### 2. Tensor-Native Operations
Ensure all DSL operations use PyTorch tensors:

```python
def tensor_operations_dsl():
    """Collection of differentiable tensor operations for DSL"""
    
    def distance_to_point(state, target_x, target_y):
        """Euclidean distance to target point"""
        target = torch.tensor([target_x, target_y], device=state.device)
        return torch.norm(state[:2] - target)
    
    def angular_difference(current_angle, target_angle):
        """Shortest angular distance between two angles"""
        diff = torch.abs(current_angle - target_angle)
        return torch.min(diff, 2*torch.pi - diff)
    
    def soft_constraint(value, threshold, penalty_weight=1.0):
        """Soft constraint that penalizes when value > threshold"""
        violation = torch.relu(value - threshold)
        return penalty_weight * violation
    
    def exponential_barrier(value, threshold, steepness=5.0):
        """Exponential barrier function for inequality constraints"""
        return torch.exp(steepness * (value - threshold))
    
    return {
        'distance_to_point': distance_to_point,
        'angular_difference': angular_difference,
        'soft_constraint': soft_constraint,
        'exponential_barrier': exponential_barrier
    }
```

### 3. Dynamic DSL Programs
Support runtime modification of DSL programs:

```python
class DynamicDSLProgram:
    def __init__(self):
        self.instructions = []
        self.weights = {}  # instruction-specific weights
    
    def add_goal(self, x, y, weight=1.0):
        """Add goal instruction with weight"""
        instr = DSLInstruction("goal", x, y)
        self.instructions.append(instr)
        self.weights[len(self.instructions)-1] = weight
    
    def add_obstacle(self, x, y, radius, weight=10.0):
        """Add obstacle avoidance instruction"""
        instr = DSLInstruction("avoid_obstacle", x, y, radius)
        self.instructions.append(instr)
        self.weights[len(self.instructions)-1] = weight
    
    def remove_instruction(self, index):
        """Remove instruction by index"""
        if 0 <= index < len(self.instructions):
            del self.instructions[index]
            del self.weights[index]
    
    def evaluate_weighted(self, latent, state):
        """Evaluate with per-instruction weights"""
        total_loss = torch.tensor(0.0, device=state.device)
        
        for i, instr in enumerate(self.instructions):
            loss = self.evaluate_instruction(instr, latent, state)
            weight = self.weights.get(i, 1.0)
            total_loss += weight * loss
            
        return total_loss
```

### 4. Hierarchical DSL Structure
Support nested and hierarchical constructs:

```python
def hierarchical_dsl_example():
    """Example of hierarchical DSL with nested behaviors"""
    
    # High-level behavior: navigate to goal while avoiding obstacles
    navigation_program = [
        ("sequence", [
            ("goal", 5.0, 5.0),
            ("avoid_obstacles", [
                ("avoid_obstacle", 2.0, 2.0, 0.5),
                ("avoid_obstacle", 4.0, 3.0, 0.7)
            ])
        ])
    ]
    
    def evaluate_hierarchical(program, state):
        """Evaluate hierarchical DSL program"""
        total_loss = torch.tensor(0.0, device=state.device)
        
        for operation, args in program:
            if operation == "sequence":
                # Execute all sub-instructions in sequence
                for sub_instr in args:
                    loss = evaluate_hierarchical([sub_instr], state)
                    total_loss += loss
            elif operation == "avoid_obstacles":
                # Combine multiple obstacle avoidance constraints
                obstacle_loss = torch.tensor(0.0, device=state.device)
                for obs_instr in args:
                    obs_loss = evaluate_hierarchical([obs_instr], state)
                    obstacle_loss += obs_loss
                total_loss += obstacle_loss
            else:
                # Base case: evaluate primitive instruction
                loss = evaluate_dsl(None, state)  # simplified
                total_loss += loss
        
        return total_loss
```

## Acceptance Criteria
- [ ] DSL parser correctly handles symbolic instructions
- [ ] All DSL operations are fully differentiable (gradients flow through)
- [ ] Integration with main training loop preserves end-to-end optimization
- [ ] Conditional and loop constructs work with differentiable operations
- [ ] Dynamic program modification supported at runtime
- [ ] Hierarchical program structures evaluated correctly
- [ ] Performance: DSL evaluation adds <10% overhead to forward pass
- [ ] Memory usage scales linearly with number of instructions

### Meta-Learning DSL Programs
Learn to generate DSL programs automatically:

```python
class DSLMetaLearner(nn.Module):
    def __init__(self, state_dim, max_instructions=10):
        super().__init__()
        self.state_dim = state_dim
        self.max_instructions = max_instructions
        
        # Neural network that outputs DSL program
        self.program_generator = nn.Sequential(
            nn.Linear(state_dim, 128),
            nn.ReLU(),
            nn.Linear(128, max_instructions * 4)  # operation + 3 args per instruction
        )
    
    def forward(self, state):
        """Generate DSL program conditioned on current state"""
        raw_output = self.program_generator(state)
        # Parse output into DSL instructions
        instructions = []
        for i in range(self.max_instructions):
            instr_params = raw_output[i*4:(i+1)*4]
            # Use softmax to select operation type
            # Use continuous values for arguments
            instructions.append(self.parse_instruction(instr_params))
        return instructions
```

### DSL Program Optimization
Optimize DSL programs via gradient descent:

```python
def optimize_dsl_program(initial_program, training_data, num_epochs=100):
    """
    Optimize DSL program parameters to minimize loss on training data
    """
    # Convert program parameters to learnable tensors
    program_params = []
    for instr in initial_program:
        if instr.operation == "goal":
            # Make goal coordinates learnable
            goal_params = torch.tensor(instr.args, requires_grad=True)
            program_params.append(goal_params)
    
    optimizer = torch.optim.Adam(program_params, lr=0.01)
    
    for epoch in range(num_epochs):
        total_loss = torch.tensor(0.0)
        
        for state, target_behavior in training_data:
            # Evaluate current program
            program_loss = evaluate_parameterized_dsl(program_params, state)
            # Compare with target behavior
            behavior_loss = F.mse_loss(program_loss, target_behavior)
            total_loss += behavior_loss
        
        # Update program parameters
        optimizer.zero_grad()
        total_loss.backward()
        optimizer.step()
    
    return program_params
```

## Integration Notes
The DSL system will be integrated throughout the codebase:

```python
# Instantiate DSL program
dsl_program = DSLProgram([
    ("goal", 5.0, 5.0),
    ("stay_inside", 0.0, 10.0, 0.0, 10.0),
    ("avoid_obstacle", 3.0, 3.0, 1.0)
])

# In main training loop:
for step in range(num_steps):
    # Get current state from simulation
    current_state = get_current_state()
    
    # Evaluate DSL constraints
    dsl_loss = dsl_program.evaluate(latent, current_state)
    
    # Combine with other losses
    total_loss = model_loss + jepa_loss + 0.1 * dsl_loss
    
    # Optimize
    optimizer.zero_grad()
    total_loss.backward()
    optimizer.step()

# For action selection:
action = active_inference_with_dsl(current_state, dynamics_model)
```

This differentiable DSL enables **principled injection of high-level knowledge** into the learning process while maintaining full gradient flow, allowing symbolic reasoning to guide neural network optimization.
